tests/integration/util/kind.go

```diff
@@ -0,0 +1,217 @@
+}
+
+func CreateKindClusterFromConfig(t *testing.T, config kind.Cluster) string {
+	kubeconfig := kindGetKubeconfigPath(t)
+	configPath := writeKindClusterConfig(t, config)
+	t.Logf("creating kind cluster %s", config.Name)
+	out, err := exec.Command(
+		"kind", "create", "cluster",
+		"--config", configPath,
+		"--kubeconfig", kubeconfig,
+	).CombinedOutput()
+	if err != nil {
+		t.Logf("output: %s", out)
+		t.Fatalf("failed to create kind cluster: %s", err)
+	}
+	t.Logf("created kind cluster %s", config.Name)
+	waitForDefaultServiceAccount(t, kubeconfig, 30*time.Second)
+	return kubeconfig
+}
+
+func DeleteKindCluster(t *testing.T, name string) {
+	t.Logf("deleting kind cluster %s", name)
+	kubeconfig := kindGetKubeconfigPath(t)
+	out, err := exec.Command("kind", "delete", "cluster", "--name", name, "--kubeconfig", kubeconfig).CombinedOutput()
+	if err != nil {
+		t.Logf("output: %s", out)
+		t.Logf("WARN: failed to delete kind cluster: %s", err)
+	}
+	t.Logf("deleted kind cluster %s", name)
+}
+
+func DeferCleanupKindCluster(t *testing.T, name string) {
+	if os.Getenv("DEBUG") != "" {
+		return
+	}
+	t.Cleanup(func() { DeleteKindCluster(t, name) })
+}
+
+func KindGetExposedPort(t *testing.T, name string, containerPort string) string {
+	nodes := kindListNodes(t, name)
+	for _, node := range nodes {
+		p := kindNodeGetExposedPorts(t, node)
+		if v := p[containerPort]; v != "" {
+			return v
+		}
+	}
+	t.Fatalf("failed to find exposed port for container port %s", containerPort)
+	return ""
+}
+
+func NewKindClusterConfig(t *testing.T, name string, opts *KindClusterOptions) kind.Cluster {
+	config := kind.Cluster{
+		APIVersion: "kind.x-k8s.io/v1alpha4",
+		Kind:       "Cluster",
+		Name:       name,
+		Networking: kind.Networking{
+			// same as k0s
+			PodSubnet:     "10.244.0.0/16",
+			ServiceSubnet: "10.96.0.0/12",
+		},
+	}
+	numControllerNodes := 1
+	numWorkerNodes := 0
+	portMappings := []kind.PortMapping{
+		{
+			ContainerPort: 30000,
+		},
+	}
+	if opts != nil {
+		if opts.NumControlPlaneNodes > 0 {
+			numControllerNodes = opts.NumControlPlaneNodes
+		}
+		numWorkerNodes = opts.NumWorkerNodes
+		for _, port := range opts.ExposedPorts {
+			portMappings = append(portMappings, kind.PortMapping{
+				ContainerPort: port,
+			})
+		}
+	}
+	for i := 0; i < numControllerNodes; i++ {
+		node := kind.Node{
+			Role: "control-plane",
+		}
+		if numWorkerNodes == 0 && i == 0 {
+			node.ExtraPortMappings = portMappings
+		}
+		config.Nodes = append(config.Nodes, node)
+	}
+	for i := 0; i < numWorkerNodes; i++ {
+		node := kind.Node{
+			Role: "worker",
+		}
+		if i == 0 {
+			node.ExtraPortMappings = portMappings
+		}
+		config.Nodes = append(config.Nodes, node)
+	}
+	return config
+}
+
+func kindGetKubeconfigPath(t *testing.T) string {
+	return filepath.Join("/tmp", t.Name(), "kubeconfig")
+}
+
+func writeKindClusterConfig(t *testing.T, config kind.Cluster) string {
+	f, err := os.CreateTemp("", "kind-config-*.yaml")
```

Comment:
nit: we should create temp files in `t.TempDir`

Reply:
done